### DISCUSSION TITLE
Status of Attributes?

### DISCUSSION THREAD

**Discussion by aVersionOfReality** [bnpr/Malt#142] [Q&A]
What is the status of accessing Attributes from objects or object data? In Cycles/Eevee, we can access vertex colors, but also custom properties. I see that Malt has a node for vertex colors. Is it possible to access other things right now, either as a node or code? Or does that require further development?

How about Object Coordinates from any object, like the Texture Coordinate node in Cycles if you select an Object in the drop down?

**Comment by pragma37** [comment]
There's nothing like the object attributes at the moment. You can use multiple materials with the same node group and override only what you need.

For object coordinates, you'll have to use drivers.

**Reply by aVersionOfReality** [reply]
Wanted to bump this up again now that some time has passed. I think getting geometry Attributes working soon would be very helpful. There's a lot going on now with Geometry nodes being able to output arbitrary attributes that can then be brought into the shader and do all sorts of useful things. It goes far beyond what we can do with Drivers, since many of these Attributes have different values per vertex. It would be great if Malt could support these the way Eevee and Cycles do with their Attribute node. We're seeing a lot of people asking about this feature in various places.

**Reply by pragma37** [reply]
Supporting Geometry Attributes would require a lot of work. I'm not even sure if the data is actually accessible from addons.
Anyway, couldn't you just store that same info into the mesh Vertex Colors or UVs to get the same results?

**Reply by aVersionOfReality** [reply]
I believe we can use UVs or Vertex Colors in pretty much all the cases I've thought of so far. In Geometry Nodes, we can always output whatever other attribute to a Vertex Color or UV. It may cause some strange workflow clutter though, like needing to run a Geonodes group on every object. And UVs and Vertex Colors are both Face Corners, whereas in Geonodes we could be working with Vertex, Face, or Face Corner data. We can convert Vertex and Face to Face Corner, but that seems inefficient. Also may be clamping/number range issues?

So I guess this is mostly a workflow clutter and user experience issue, not a hard block on doing what I want to do (pass custom Normals from Geonodes to the shader.)

I'm not clear though on what the difference is under the hood between these and other attributes. We can use Vertex Colors, which are a vec3, but we can't use a generic vec3 attribute? Or a float attribute? Aren't these all just different sorts of data in a per vertex list? Is there some sort of way UVs and Vertex Colors are different in openGL/glsl or something?

Thanks!

**Reply by pragma37** [reply]
> And UVs and Vertex Colors are both Face Corners, whereas in Geonodes we could be working with Vertex, Face, or Face Corner data. We can convert Vertex and Face to Face Corner, but that seems inefficient.

Not really, there's no such thing as Face Corners on the GPU side, if a vertex has "per-face" data, then you actually have 2 (or more) vertices. 

> Also may be clamping/number range issues?

There would be for vertex colors but not for uvs.

> I'm not clear though on what the difference is under the hood between these and other attributes. We can use Vertex Colors, which are a vec3, but we can't use a generic vec3 attribute? Or a float attribute? Aren't these all just different sorts of data in a per vertex list? Is there some sort of way UVs and Vertex Colors are different in openGL/glsl or something?

OpenGL has no built-in concept of mesh UVs or Vertex Colors. The issue is that shaders must be compiled for a specific vertex data layout.
Malt uses the same vertex layout for all shaders so the system is simpler overall, and to avoid having to compile the same material multiple times for different mesh layouts. And since there's a limit of 16 vertex attributes, with the current system we can't just keep adding attributes carelessly.

In Blender, vertex colors are 8 bits per channel, but Malt doesn't have such a limitation. I've discussed this with Kolupsy today and I think a somewhat simpler workaround could be to add the option to override vertex color channels with mesh attributes at mesh loading time.
So on the Malt side itself there would be no need for any changes. The main limitation is that only vec4 face-corner attribute types would be supported and that, at max, 4 vertex colors are supported.

**Reply by aVersionOfReality** [reply]
Ah, alright. Thanks for the explanation. I knew that at low levels face corners don't exist, but I wasn't sure what level we're dealing with in regards to this. I'll talk with Kolupsy about it. I'm pretty sure 4 vertex colors would cover all the scenarios I can think of right now, so it should be fine.

**Comment by pragma37** [comment]
I quickly put together a *Vertex Color Overrides* feature. (f7b9641f98ee503df1f2f2c27997fa2ab0b726cd)

It only works with Face Corners attributes of type Color:
![imagen](https://user-images.githubusercontent.com/36610999/158270339-2c84a7f1-c9d0-4204-b87f-eabe750555c4.png)

**Reply by aVersionOfReality** [reply]
Ah excellent! That looks like it'll be convenient.

**Reply by aVersionOfReality** [reply]
Sorry to bump an old thread, but I wanted to revisit this. Now that I'm actively working with Malt, I'm finding I could definitely use more than 4 Vertex Colors to get all my Geonodes data in. I know we've talked about this a couple times in the past in various places, but my recollection is that we're limited to 4 vertex colors, but that could be increased if things were decreased in other areas? Can you tell me what the tradeoff is? UVs?

I know we have 4 Vertex Colors and 4 UVs at the moment. I am only using one UV map in the shader, should I could put more data into the other 3 UVs. From a geonodes perspective this isn't really any different than using the vertex colors. So if the tradeoff would be less UVs, then it doesn't matter.

**Reply by pragma37** [reply]
In the past, we were limited because we supported pre-computed tangents for each UV channel, but now only the first UV channel  supports them, so there's some room for extra vertex attributes.
Adding support for extra vertex colors should be easy, but it may be worth it to rethink the way vertex attributes are currently handled instead.

**Reply by pragma37** [reply]
In any case, yes, you can put whatever you want into the UV data.

**Reply by aVersionOfReality** [reply]
I could easily use 6-8 if that was an option. Its not mandatory, but being able to set more things in Geonodes does make working a lot easier, reduces object and material count, etc. Currently I put a geonodes group on every object that defines a lot of attributes that are used. If I could, i'd be using all of these:
- Color1 (lit color)
- Color2 (shaded color, or perhaps Line color))
- Custom Normal
- Custom Tangent
- Custom Normal 2 (More blurred, not always used)
- Line Parameters (VC ID mask, normal scale, depth scale, overall scale)
- Masks
